### PR TITLE
Add operations: `BitXor`, `BitOr`, and `Not` to MiniRust

### PR DESCRIPTION
**Pull request by bifbof**
Part of https://github.com/minirust/minirust/issues/186
Should be pretty straight for the most part.

~~But for the shift operation I am unsure if i created the correct semantics.~~
~~The [rustc doc](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.BinOp.html#variant.Shl) states:~~
```
Shl: The << operator (shift left)

The offset is truncated to the size of the first operand and made unsigned before shifting.
```
~~For made unsigned I see two possibilities:~~
~~1. C-style: interpret the bits as unsigned integers now (e.g. `-1i8` &rarr; `255u8`)~~
~~2. Throw Error (but not UB?) if trying to shift with negative value.~~
~~So far the 2nd version is implemented by `libspecr` (via BigInt).~~
~~

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks!
Please restrict this to BitXor, BitOr, Not so that we can discuss the shifts separately.

File: spec/lang/syntax.md [hunk 1]
```diff
@@ -100,6 +100,8 @@ pub enum Constant {
 pub enum UnOpInt {
     /// Negate an integer value arithmetically (`x` becomes `-x`).
     Neg,
+    /// Bitwise-not integer value
```
Comment (RalfJung) [inline spec/lang/syntax.md:103]:
```suggestion
    /// Bitwise-negate an integer value
```

File: tooling/minimize/src/rvalue.rs [hunk 2]
```diff
@@ -20,60 +20,54 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
...
-                            operand: GcCow::new(operand),
-                        }
-                    }
-                },
+                use smir::UnOp::*;
+                let operator = match (unop, ty) {
+                    (Neg, Type::Int(_)) => UnOp::Int(UnOpInt::Neg),
+                    (Not, Type::Int(_)) => UnOp::Int(UnOpInt::Not),
+                    (Not, Type::Bool) => UnOp::Bool(UnOpBool::Not),
+                    (op, _) => rs::span_bug!(span, "UnOp {op:?} called with unsupported type."),
```
Comment (RalfJung) [inline tooling/minimize/src/rvalue.rs:67]:
```suggestion
                    (op, ty) => rs::span_bug!(span, "UnOp {op:?} called with unsupported type {ty}"),
```

File: tooling/minimize/src/rvalue.rs [hunk 3]
```diff
@@ -20,60 +20,54 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
...
-                            ty => rs::span_bug!(span, "BitAnd not supported for: {ty:?}"),
-                        },
-                    x => rs::span_bug!(span, "Binary Op not supported: {x:?}"),
+                    // all bool ops
+                    (BitAnd, Type::Bool) => BinOp::Bool(BinOpBool::BitAnd),
+                    (BitOr, Type::Bool) => BinOp::Bool(BinOpBool::BitOr),
+                    (BitXor, Type::Bool) => BinOp::Bool(BinOpBool::BitXor),
+
+                    (op, ty) =>
+                        rs::span_bug!(span, "Binary Op {op:?} not supported for type {ty:?}"),
```
Comment (RalfJung) [inline tooling/minimize/src/rvalue.rs:52]:
```suggestion
                        rs::span_bug!(span, "Binary Op {op:?} not supported for type {ty}"),
```
The debug rendering of types can be quite verbose

File: tooling/minimize/tests/pass/ops.rs [hunk 4]
```diff
@@ -5,12 +5,17 @@ fn black_box<T>(t: T) -> T { t }
 
 fn main() {
     print(-black_box(-42));
+    print(!black_box(-43));
```
Comment (RalfJung) [inline tooling/minimize/tests/pass/ops.rs:8]: Please don't add more `print` tests -- it's so hard to figure out which line of output corresponds to which part of the test.

Instead, use `assert!`.

Comment (RalfJung) [inline tooling/minimize/tests/pass/ops.rs:8]: Please also check an unsigned number, like `!1u32` becoming `(u32::MAX-1)`

File: tooling/minimize/src/rvalue.rs (hunk 3 above)
Comment (bifbof) [reply tooling/minimize/src/rvalue.rs]: Sadly I cannot do this as types do not implement `std::fmt::Display`. So for now I dropped printing the types at all.

File: tooling/minimize/tests/pass/ops.rs (hunk 4 above)
Comment (bifbof) [reply tooling/minimize/tests/pass/ops.rs]: Added `assert!(!black_box(i32::MIN) == i32::MAX);` as well for good measure :)

**Comment by bifbof** [comment]
I removed the shift operations.

File: tooling/minimize/src/rvalue.rs (hunk 3 above)
Comment (RalfJung) [reply tooling/minimize/src/rvalue.rs]: Rust types implement `Display`... can we print the Rust type instead?

**Comment by RalfJung** [comment]
> I removed the shift operations.

Thanks! Please also remove them from the PR title and commit message.

**Comment by RalfJung** [comment]
> [BitXor, BitOr, and Ǹot to MiniRust](https://github.com/minirust/minirust/pull/187/commits/5f447cdf289914f02ba7593613261d9a93cc4953)

Your commit message has a strange `Ǹ` character in it still. ;)

**Comment by bifbof** [comment]
> > [BitXor, BitOr, and Ǹot to MiniRust](https://github.com/minirust/minirust/pull/187/commits/5f447cdf289914f02ba7593613261d9a93cc4953)
> 
> Your commit message has a strange `Ǹ` character in it still. ;)

oh man, thanks for the reminder.

**Comment by RalfJung** [comment]
Looks good, thanks!

**Pull request merged**